packages/devtools_app/lib/src/charts/flutter/flame_chart.dart

```diff
@@ -153,39 +170,126 @@ class _ScrollingFlameChartRowState extends State<ScrollingFlameChartRow>
           : Stack(
               children: [
                 Container(
                   height: rowHeightWithPadding,
                   width: widget.width,
                 ),
-                for (var node in widget.nodes)
-                  node.buildWidget(node.data == widget.selected),
+                ...rowNodesInViewport(),
               ],
             ),
     );
   }
 
-// TODO(kenz): consider using this method when we have a larger data set to
-// test with.
-//  List<FlameChartNode> rowNodesInViewport(
-//    List<FlameChartNode> nodes,
-//    BoxConstraints constraints,
-//  ) {
-//    // TODO(kenz): Use binary search method we use in html full timeline here.
-//    final nodesInViewport = <FlameChartNode>[];
-//    for (var node in nodes) {
-//      final horizontalScrollOffset = scrollController.hasClients
-//          ? scrollController.offset
-//          : scrollController.initialScrollOffset;
-//      final fitsHorizontally = node.rect.right >= horizontalScrollOffset &&
-//          node.rect.left - horizontalScrollOffset <= constraints.maxWidth;
-//      if (fitsHorizontally) {
-//        nodesInViewport.add(node);
-//      }
-//    }
-//    return nodesInViewport;
-//  }
+  List<Widget> rowNodesInViewport() {
+    final nodes = widget.nodes;
+    final nodesInViewport = <Widget>[];
+    final startNodeIndex = findFirstIndexInView(nodes);
+    if (startNodeIndex != -1) {
+      for (int i = startNodeIndex; i < nodes.length; i++) {
+        final node = nodes[i];
+        if (!nodeFitsInViewport(node)) {
+          break;
+        }
+        nodesInViewport.add(node.buildWidget(node.data == widget.selected));
+      }
+    }
+    lastStartNodeIndexInViewport = startNodeIndex;
+    return nodesInViewport;
+  }
+
+  int findFirstIndexInView(List<FlameChartNode> nodes) {
+    // If we know the previous start node index, start there to find the current
+    // start node index.
+    if (lastStartNodeIndexInViewport != -1) {
+      var index = lastStartNodeIndexInViewport;
+      while (index >= 0 && index < nodes.length) {
+        final node = nodes[index];
+        if (nodeFitsInViewport(node)) {
+          if (index > 0 && nodeFitsInViewport(nodes[index - 1])) {
+            // Since the previous node also fits in the viewport, keep looking
+            // left for the first fitting index.
+            index--;
+          } else {
+            // [index] is the first fitting index.
+            break;
+          }
+        } else {
+          index++;
+        }
+      }
+      // No nodes in this row fit within the viewport.
+      if (index == nodes.length) return -1;
+      return index;
+    }
+    // If we don't know the previous start node index, binary search to find the
+    // first fitting index (if one exists).
+    else {
+      final index = lowerBound(
+        nodes,
+        // Dummy node that has the left edge of the visible viewport.
+        FlameChartNode(
```

Comment:
lowerBound returns the index of the first item that does not compare as less than the given value (dummy node). With the given comparator, this will return the farthest left node whose left edge is not less than the dummy node (which is set on the edge of the viewport).

https://api.flutter.dev/flutter/package-collection_collection/lowerBound.html